### PR TITLE
Unify DefaultBlockWorkerClient with AbstractClient

### PR DESCRIPTION
**Pull request by YangchenYe323**
### What changes are proposed in this pull request?
Make `DefaultBlockWorkerClient` extends from `AbstractClient`

### Why are the changes needed?
- Address a todo issue in the codebase
- Reduce duplicated code
- Avoid doing the expensive connection when `DefaultBlockWorkerClient` is constructed, take advantage of the lazy connection mechanism of `AbstractClient`
- Improve the flexibility of `AbstractClient`

### Does this PR introduce any user facing changes?
No

### CODE REVIEW

**Comment by YangchenYe323** [comment]
In order for `DefaultBlockWorkerClient` to inherit from `AbstractClient`, three issues have to be addressed:
1. The retry scheme in `AbstractClient` is extended to separate connection and actual rpc calls, so that `DefaultBlockWorkerClient` can use a normal retry for connection but not use retry for actual rpcs, as retrying read & write might be undesirable.
2. `DefaultBlockWorkerClient` might not work with an `InetSocketAddress` as it might use a domain socket instead. Therefore, `AbstractClient` is modified to expose a `getGrpcServerAddress` method used to return whatever server address the client connects to. `DefaultBlockWorkerClient` overrides this method to return the address it needs for connection.
3. `AbstractClient` now can be extended to construct multiple channels corresponding to different `GrpcNetworkGroup`, so that `DefaultBlockWorker` can construct two channels, one for streaming and one for normal rpc, for its use.

File: core/client/fs/src/main/java/alluxio/client/block/stream/BlockWorkerClient.java [hunk 1]
```diff
@@ -48,20 +48,15 @@ class Factory {
...
-      try {
-        return new DefaultBlockWorkerClient(userState, address, alluxioConf);
-      } catch (Exception e) {
-        throw new IOException(
-            String.format("Failed to connect to block worker (%s)", address), e);
-      }
+    public static BlockWorkerClient create(ClientContext context,
+                                           GrpcServerAddress address,
+                                           AlluxioConfiguration alluxioConf) {
+      return new DefaultBlockWorkerClient(context, address, alluxioConf);
```
Comment (YangchenYe323) [inline core/client/fs/src/main/java/alluxio/client/block/stream/BlockWorkerClient.java:59]: Since we get rid of the connecting logic in the constructor of `DefaultBlockWorker`, it's impossible to have any exception now.

File: core/client/fs/src/main/java/alluxio/client/block/stream/BlockWorkerClient.java [hunk 2]
```diff
@@ -91,7 +87,8 @@ public static BlockWorkerClient create(UserState userState, GrpcServerAddress ad
    * @param responseObserver the stream observer for the server response
    * @return the stream observer for the client request
    */
-  StreamObserver<ReadRequest> readBlock(StreamObserver<ReadResponse> responseObserver);
+  StreamObserver<ReadRequest> readBlock(StreamObserver<ReadResponse> responseObserver)
+          throws IOException;
```
Comment (YangchenYe323) [inline core/client/fs/src/main/java/alluxio/client/block/stream/BlockWorkerClient.java:91]: The `BlockWorker` interface is modified to throw `IOException` that comes with using `AbstractClient#retryRPC`

File: core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java [hunk 3]
```diff
@@ -73,61 +72,43 @@ public class DefaultBlockWorkerClient implements BlockWorkerClient {
...
-      throw lastException;
-    }
-    mStreamingAsyncStub = BlockWorkerGrpc.newStub(mStreamingChannel);
-    mRpcBlockingStub = BlockWorkerGrpc.newBlockingStub(mRpcChannel);
-    mRpcAsyncStub = BlockWorkerGrpc.newStub(mRpcChannel);
+  public DefaultBlockWorkerClient(ClientContext context, GrpcServerAddress address,
+                                  AlluxioConfiguration alluxioConf) {
+    // we don't use InetSocketAddress
+    // we use default retry policy for connection and no retry for actual rpc
+    super(context, null, RetryUtils::defaultClientRetry, RetryUtils::oneshotRetryPolicy);
```
Comment (YangchenYe323) [inline core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java:93]: As commented, we retry normally for connection but don't retry for real rpc calls.

File: core/client/fs/src/main/java/alluxio/client/block/stream/GrpcBlockingStream.java [hunk 4]
```diff
@@ -308,6 +307,18 @@ private String formatErrorMessage(String format, Object... args) {
...
+   *
+   * @param <ReqT> Rpc Request Type
+   * @param <ResT> Rpc Response Type
+   */
+  @FunctionalInterface
+  public interface StreamRpcFunction<ReqT, ResT> {
+    @SuppressWarnings("checkstyle:MissingJavadocMethod")
+    StreamObserver<ReqT> apply(StreamObserver<ResT> t) throws IOException;
+  }
+
```
Comment (YangchenYe323) [inline core/client/fs/src/main/java/alluxio/client/block/stream/GrpcBlockingStream.java:321]: Since our `BlockWorkerClient` methods throw exception, we can't use java `Function`, so we have to create a functional interface here

File: core/common/src/main/java/alluxio/AbstractClient.java [hunk 5]
```diff
@@ -152,6 +180,13 @@ protected void checkVersion(long clientVersion) throws IOException {
     }
   }
 
+  /**
+   * @return If the client should skip version check service
+   */
+  protected boolean skipVersionCheck() {
+    return false;
+  }
```
Comment (YangchenYe323) [inline core/common/src/main/java/alluxio/AbstractClient.java:188]: `BlockWorkerClient` doesn't check versions as alluxio workers don't hold version service. This provides a switch for inheriting classes to skip version checks

File: core/common/src/main/java/alluxio/AbstractClient.java [hunk 6]
```diff
@@ -160,6 +195,10 @@ protected void afterConnect() throws IOException {
     // Empty implementation.
   }
 
+  protected Set<GrpcNetworkGroup> getAdditionalNetworkGroups() {
+    return ImmutableSet.of();
+  }
```
Comment (YangchenYe323) [inline core/common/src/main/java/alluxio/AbstractClient.java:200]: This provides a control handle for inheriting classes to designate additional channels they need.

File: core/common/src/main/java/alluxio/AbstractClient.java [hunk 7]
```diff
@@ -311,6 +379,26 @@ public synchronized InetSocketAddress getAddress() throws UnavailableException {
...
+   * @throws UnavailableException when no address can be determined
+   */
+  protected synchronized GrpcServerAddress getGrpcServerAddress() throws UnavailableException {
+    mAddress = getAddress();
+    if (mAddress != null) {
+      return GrpcServerAddress.create(mAddress);
+    } else {
+      throw new UnavailableException("Cannot Determine Server Address");
+    }
+  }
```
Comment (YangchenYe323) [inline core/common/src/main/java/alluxio/AbstractClient.java:400]: As commented, not all clients talk to a remote service through internet, this method exposes a way for inheriting classes to define their own address to connect to.

File: core/common/src/main/java/alluxio/retry/RetryUtils.java [hunk 8]
```diff
@@ -99,6 +99,13 @@ public static RetryPolicy defaultWorkerMasterClientRetry() {
         .build();
   }
 
+  /**
+   * @return one shot retry policy that just try once
+   */
+  public static RetryPolicy oneshotRetryPolicy() {
+    return new CountingRetry(1);
+  }
```
Comment (YangchenYe323) [inline core/common/src/main/java/alluxio/retry/RetryUtils.java:107]: I wonder if there is an existing `RetryPolicy` that just does no retry?

File: core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java [hunk 9]
```diff
@@ -73,182 +72,213 @@ public class DefaultBlockWorkerClient implements BlockWorkerClient {
...
-      throw lastException;
-    }
-    mStreamingAsyncStub = BlockWorkerGrpc.newStub(mStreamingChannel);
-    mRpcBlockingStub = BlockWorkerGrpc.newBlockingStub(mRpcChannel);
-    mRpcAsyncStub = BlockWorkerGrpc.newStub(mRpcChannel);
+  public DefaultBlockWorkerClient(ClientContext context, GrpcServerAddress address,
+                                  AlluxioConfiguration alluxioConf) {
+    // we don't use InetSocketAddress
+    // we use default retry policy for connection and no retry for actual rpc
+    super(context, null, RetryUtils::oneshotRetryPolicy, RetryUtils::oneshotRetryPolicy);
```
[the remark this answers is not in the document]
Comment (YangchenYe323) [reply core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java:93]: After some trial and error, I switched to no retry for both connection and actual rpc. This should fix the test failure. Also, we might not need the separation of connection & rpc RetryPolicy at all.

File: core/common/src/main/java/alluxio/retry/RetryUtils.java (hunk 8 above)
Comment (jja725) [reply core/common/src/main/java/alluxio/retry/RetryUtils.java:107]: you can input CountingRetry(0) for this purpose

File: core/client/fs/src/main/java/alluxio/client/block/stream/BlockWorkerClient.java (hunk 2 above)
Comment (jja725) [reply core/client/fs/src/main/java/alluxio/client/block/stream/BlockWorkerClient.java:91]: feel like we should update the retryRPC signature so we just throw runtime exception

File: core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java [hunk 10]
```diff
@@ -73,182 +72,219 @@ public class DefaultBlockWorkerClient implements BlockWorkerClient {
...
+    }
+
+    if (isConnected()) {
+      return mStreamingChannel != null && mStreamingChannel.isHealthy()
+              && mRpcChannel != null && mRpcChannel.isHealthy();
+    }
+
+    // client is not connected yet, assume it is healthy
+    return true;
   }
```
Comment (YangchenYe323) [inline core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java:140]: I'm a little bit uncomfortable about this check. It's subject to the risk of false negative: it might falsely think some unhealthy client is healthy because `retryRPC` disconnects on rpc failure, and hence a client that has failed several rpc calls is indistinguishable from a fresh client that has not connected yet.

**Comment by YangchenYe323** [comment]
@beinan Do you mind taking a look at this refactor? Thanks a lot!

File: core/common/src/main/java/alluxio/AbstractClient.java [hunk 11]
```diff
@@ -85,29 +96,30 @@ public abstract class AbstractClient implements Client {
...
 
   /**
-   * Creates a new client base.
+   * Creates a new client base with default retry policy supplier.
    *
    * @param context information required to connect to Alluxio
    * @param address the address
    */
-  public AbstractClient(ClientContext context, InetSocketAddress address) {
+  public AbstractClient(ClientContext context, @Nullable InetSocketAddress address) {
```
Comment (dbw9580) [inline core/common/src/main/java/alluxio/AbstractClient.java:104]: I don't think allowing nullable arguments in public constructors is a good idea. If this parameter can be null, provide an overloaded constructor without this parameter, and internally calling a private constructor that handles nullable arguments.

File: core/common/src/main/java/alluxio/AbstractClient.java [hunk 12]
```diff
@@ -54,13 +61,17 @@
 public abstract class AbstractClient implements Client {
   private static final Logger LOG = LoggerFactory.getLogger(AbstractClient.class);
 
+  // RetryPolicy Supplier for business rpc calls
   private final Supplier<RetryPolicy> mRetryPolicySupplier;
 
   protected InetSocketAddress mAddress;
```
Comment (dbw9580) [inline core/common/src/main/java/alluxio/AbstractClient.java:67]: Can we make this `Optional<InetSocketAddress>`?

Comment (dbw9580) [reply core/common/src/main/java/alluxio/AbstractClient.java:67]: On closer look, why is this nullable?

File: core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java [hunk 13]
```diff
@@ -73,182 +72,219 @@ public class DefaultBlockWorkerClient implements BlockWorkerClient {
...
 
+  @Override
+  protected void beforeConnect() {}
+
+  @Override
+  protected void afterConnect() {
+    // these are alias to the channel built in parent class
+    mStreamingChannel = Preconditions.checkNotNull(
+            mAdditionalChannels.get(GrpcNetworkGroup.STREAMING));
```
Comment (dbw9580) [inline core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java:112]: The channels are constructed in the super class, but re-assigned/aliased to a field in a child class - this is very a confusing and convoluted solution.

File: core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java [hunk 14]
```diff
@@ -73,182 +72,219 @@ public class DefaultBlockWorkerClient implements BlockWorkerClient {
...
   public boolean isHealthy() {
-    return !isShutdown() && mStreamingChannel.isHealthy() && mRpcChannel.isHealthy();
+    if (isShutdown()) {
+      // client is shut down
+      return false;
+    }
+
+    if (isConnected()) {
+      return mStreamingChannel != null && mStreamingChannel.isHealthy()
+              && mRpcChannel != null && mRpcChannel.isHealthy();
```
Comment (dbw9580) [inline core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java:135]: How come the channels can be null, if you have `Preconditions.checkNotNull`ed them?

**Review by dbw9580 — CHANGES_REQUESTED** [review]

File: core/common/src/main/java/alluxio/AbstractClient.java (hunk 5 above)
Comment (YangchenYe323) [reply core/common/src/main/java/alluxio/AbstractClient.java:172]: Yeah that's true. I finally realized `BlockWorkerGrpc`'s service type is actally `FILE_SYSTEM_WORKER_WORKER_SERVICE`.

Comment (YangchenYe323) [reply core/common/src/main/java/alluxio/AbstractClient.java:172]: I added this method because the original `DefaultBlockWorkerClient` doesn't check version though.

File: core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java (hunk 14 above)
Comment (YangchenYe323) [reply core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java:135]: They can't according to the current logic. It's just a safety short circuit.

File: core/common/src/main/java/alluxio/AbstractClient.java (hunk 12 above)
Comment (YangchenYe323) [reply core/common/src/main/java/alluxio/AbstractClient.java:67]: `DefaultBlockWorkerClient` doesn't have an `InetSocketAddress`. All it has is a `GrpcServerAddress`, which might contain a domain socket.

For other client like `AbstractMasterClient`, `mAddress` is not initialized until it queries the `MasterInquireClient`, so it could also be `null` at some point in the client's life time.

File: core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java (hunk 13 above)
Comment (YangchenYe323) [reply core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java:112]: We could get rid of the alias pretty easily. They're here just to be compatible with the original code.

File: core/client/fs/src/main/java/alluxio/client/block/stream/BlockWorkerClient.java (hunk 2 above)
Comment (jja725) [reply core/client/fs/src/main/java/alluxio/client/block/stream/BlockWorkerClient.java:91]: In that case, AlluxioStatusException makes more sense here and we can add more information like retryable to AlluxioStatusException

File: core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java (hunk 14 above)
Comment (jja725) [reply core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java:135]: we should remove it if it doesn't gonna happen

File: core/common/src/main/java/alluxio/AbstractClient.java [hunk 15]
```diff
@@ -85,29 +96,30 @@ public abstract class AbstractClient implements Client {
...
+  public AbstractClient(ClientContext context, @Nullable InetSocketAddress address,
       Supplier<RetryPolicy> retryPolicySupplier) {
     mAddress = address;
     mContext = Preconditions.checkNotNull(context, "context");
-    mRetryPolicySupplier = retryPolicySupplier;
+    mRetryPolicySupplier = Preconditions.checkNotNull(retryPolicySupplier);
     mServiceVersion = Constants.UNKNOWN_SERVICE_VERSION;
     mRpcThreshold = mContext.getClusterConf().getMs(PropertyKey.USER_LOGGING_THRESHOLD);
+    mAdditionalChannels = new HashMap<>();
   }
```
Comment (jja725) [inline core/common/src/main/java/alluxio/AbstractClient.java:123]: put field assignment with definition if it's not passed in

File: core/common/src/main/java/alluxio/AbstractClient.java (hunk 5 above)
Comment (jja725) [reply core/common/src/main/java/alluxio/AbstractClient.java:172]: I would give it a new name like BLOCK_WORKER_CLIENT_SERVICE. And could we just use a constant number in getServiceVersion to avoid adding a new API? your call

File: core/server/master/src/main/java/alluxio/master/file/loadmanager/LoadManager.java [hunk 16]
```diff
@@ -286,7 +287,12 @@ private boolean scheduleBatch(
     }
     LoadRequest request = buildRequest(
         batch, load.getBandWidth() / mActiveWorkers.size());
-    ListenableFuture<LoadResponse> responseFuture = workerClient.get().load(request);
+    ListenableFuture<LoadResponse> responseFuture;
+    try {
+      responseFuture = workerClient.get().load(request);
+    } catch (AlluxioStatusException e) {
+      throw e.toGrpcStatusRuntimeException();
+    }
```
Comment (dbw9580) [inline core/server/master/src/main/java/alluxio/master/file/loadmanager/LoadManager.java:295]: @rongrong we are refactoring `DefaultBlockWorkerClient` to throw checked exceptions. Do you know what's the appropriate way to handle the checked exceptions here?

**Comment by YangchenYe323** [comment]
@dbw9580 Cleaned up a bit and addressed your concerns. Would you like to take another look? Thanks!

File: core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java [hunk 17]
```diff
@@ -38,207 +40,241 @@
...
-        if (mTracker != null) {
-          mTracker.close(this);
-        }
-      });
+  protected void shutdownChannelsIfCreated() {
+    super.shutdownChannelsIfCreated();
+    if (mStreamingChannel != null) {
+      mStreamingChannel.shutdown();
+      mStreamingChannel = null;
     }
```
Comment (YangchenYe323) [inline core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java:134]: This `null` check is necessary because this method will be called on partial failures, so we can't be sure which channels are established and which are not.

File: core/client/fs/src/main/java/alluxio/client/block/stream/BlockWorkerClientPool.java [hunk 18]
```diff
@@ -48,25 +48,37 @@ public final class BlockWorkerClientPool extends DynamicResourcePool<BlockWorker
...
   private final AlluxioConfiguration mConf;
 
+  /**
+   * Create a block master client pool with default capacities.
+   *
+   * @param context Client context
+   * @param address Address of worker
+   * @param alluxioConf Alluxio Configuration
+   */
+  public BlockWorkerClientPool(ClientContext context,
```
Comment (dbw9580) [inline core/client/fs/src/main/java/alluxio/client/block/stream/BlockWorkerClientPool.java:58]: `alluxioConf` can be obtained from `context` so the 3rd parameter is redundant.

File: core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java [hunk 19]
```diff
@@ -38,207 +40,221 @@
...
-    UnauthenticatedException lastException = null;
-    // TODO(feng): unify worker client with AbstractClient
-    while (retryPolicy.attempt()) {
-      try {
-        // Disables channel pooling for data streaming to achieve better throughput.
-        // Channel is still reused due to client pooling.
-        mStreamingChannel = GrpcChannelBuilder.newBuilder(address, alluxioConf)
-            .setSubject(userState.getSubject())
+  public DefaultBlockWorkerClient(ClientContext context, GrpcServerAddress address,
+                                  AlluxioConfiguration alluxioConf) {
```
Comment (dbw9580) [inline core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java:79]: `alluxioConf` can be obtained from `ClientContext`, so it's redundant.

File: core/client/fs/src/main/java/alluxio/client/block/stream/GrpcBlockingStream.java [hunk 20]
```diff
@@ -75,8 +74,8 @@
    * @param bufferSize maximum number of incoming messages the buffer can hold
    * @param description description of this stream
    */
-  public GrpcBlockingStream(Function<StreamObserver<ResT>, StreamObserver<ReqT>> rpcFunc,
```
Comment (dbw9580) [inline core/client/fs/src/main/java/alluxio/client/block/stream/GrpcBlockingStream.java:78]: why was this necessary? the constructor now has checked exception.

File: core/common/src/main/java/alluxio/exception/status/AlluxioStatusException.java [hunk 21]
```diff
@@ -111,6 +111,13 @@ public StatusException toGrpcStatusException() {
     return mStatus.asException();
   }
 
+  /**
+   * @return a {@link StatusRuntimeException} for this exception
+   */
+  public StatusRuntimeException toGrpcStatusRuntimeException() {
+    return mStatus.asRuntimeException();
```
Comment (dbw9580) [inline core/common/src/main/java/alluxio/exception/status/AlluxioStatusException.java:118]: this is a good addition, but is unrelated to this PR, please open another PR.

**Review by dbw9580 — CHANGES_REQUESTED** [review]

File: core/client/fs/src/test/java/alluxio/client/file/MockFileSystemMasterClient.java [hunk 22]
```diff
@@ -194,8 +193,7 @@ public void connect() throws IOException {
   public void disconnect() {
   }
 
-  @Override
-  public SocketAddress getRemoteSockAddress() throws UnavailableException {
+  public InetSocketAddress getRemoteSockAddress() throws UnavailableException {
```
Comment (dbw9580) [inline core/client/fs/src/test/java/alluxio/client/file/MockFileSystemMasterClient.java:196]: why this change?

File: core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java [hunk 23]
```diff
@@ -38,207 +40,221 @@
...
+  protected void doConnect() throws AlluxioStatusException {
+    super.doConnect();
+
+    // build a streaming channel
+    AlluxioConfiguration conf = mContext.getClusterConf();
+    mStreamingChannel = GrpcChannelBuilder.newBuilder(mServerAddress, conf)
+            .setSubject(mContext.getSubject())
             .setNetworkGroup(GrpcNetworkGroup.STREAMING)
             .build();
-        mStreamingChannel.intercept(new StreamSerializationClientInterceptor());
```
Comment (dbw9580) [inline core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java:104]: why this is removed?

**Comment by dbw9580** [comment]
Please try your best to preserve the behavior of the client, since this is a fundamental class, a lot of things depend on it.

File: core/common/src/main/java/alluxio/AbstractClient.java [hunk 24]
```diff
@@ -268,10 +284,6 @@ public synchronized void connect() throws AlluxioStatusException {
       }
     }
     // Reaching here indicates that we did not successfully connect.
-    if (mChannel != null) {
-      mChannel.shutdown();
-    }
```
[the remark this answers is not in the document]
Comment (YangchenYe323) [reply core/common/src/main/java/alluxio/AbstractClient.java:273]: This functionality is moved into `shutdownChannelsIfCreated` method

File: core/client/fs/src/main/java/alluxio/client/block/stream/GrpcBlockingStream.java (hunk 20 above)
Comment (YangchenYe323) [reply core/client/fs/src/main/java/alluxio/client/block/stream/GrpcBlockingStream.java:78]: It should have, because the `rpcFunc` is `BlockWorkerClient#readBlock`, which throws checked exception. A `Function` can't throw checked exception in the original code.

File: core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java (hunk 23 above)
Comment (YangchenYe323) [reply core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java:104]: Added back

File: core/common/src/main/java/alluxio/AbstractClient.java [hunk 25]
```diff
@@ -438,7 +462,7 @@ protected synchronized <V> V retryRPC(RetryPolicy retryPolicy, RpcCallable<V> rp
             CommonUtils.summarizeCollection(ret), duration, mRpcThreshold);
       }
       return ret;
-    } catch (Exception e) {
+    } catch (AlluxioStatusException e) {
```
[the remark this answers is not in the document]
Comment (YangchenYe323) [reply core/common/src/main/java/alluxio/AbstractClient.java:465]: I suppose it doesn't change the behavior since `AlluxioStatusException` is the only checked exception that can be thrown.

File: core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java [hunk 26]
```diff
@@ -38,207 +40,221 @@
...
-            .setNetworkGroup(GrpcNetworkGroup.RPC)
-            .build();
-        lastException = null;
-        break;
-      } catch (StatusRuntimeException e) {
-        close();
-        throw AlluxioStatusException.fromStatusRuntimeException(e);
-      } catch (UnauthenticatedException e) {
-        close();
-        userState.relogin();
```
[the remark this answers is not in the document]
Comment (YangchenYe323) [reply core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java:117]: How about retrying on `UnauthenticatedException` unconditionally? It won't work to distinguish between initial connection and other connections, it would still break if the worker failed before the first connection is made.

File: core/client/fs/src/test/java/alluxio/client/file/MockFileSystemMasterClient.java (hunk 22 above)
Comment (YangchenYe323) [reply core/client/fs/src/test/java/alluxio/client/file/MockFileSystemMasterClient.java]: Shouldn't be there

File: core/client/fs/src/main/java/alluxio/client/block/stream/BlockWorkerClientPool.java (hunk 18 above)
Comment (YangchenYe323) [reply core/client/fs/src/main/java/alluxio/client/block/stream/BlockWorkerClientPool.java:57]: Done

File: core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java (hunk 19 above)
Comment (YangchenYe323) [reply core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java]: Fixed

File: core/common/src/main/java/alluxio/exception/status/AlluxioStatusException.java (hunk 21 above)
Comment (YangchenYe323) [reply core/common/src/main/java/alluxio/exception/status/AlluxioStatusException.java:118]: Done

**Review by dbw9580 — APPROVED** [review]
LGTM.

@tcrain @rongrong can you please take a look at this PR?

**Comment by yyongycy** [comment]
What tests have been done? How do you make sure not to break some conner cases?

**Comment by YangchenYe323** [comment]
> What tests have been done? How do you make sure not to break some conner cases?

Every test case that reads or writes from alluxio uses this class. Also tested the generated tarball against tests in ClientIO cluster.

**Review by jja725 — APPROVED** [review]
LGTM although I would prefer to throw AlluxioRuntimeException and put it in the comments. We can work on that later on and clean up exception handling at the client side.
